src/passes/CallArgumentOptimization.cpp

```diff
@@ -0,0 +1,366 @@
+          auto* operand = call->operands[i];
+          if (auto* c = operand->dynCast<Const>()) {
+            if (value.type == none) {
+              // This is the first value seen.
+              value = c->value;
+            } else if (value != c->value) {
+              // Not identical, give up
+              value.type = none;
+              break;
+            }
+          } else {
+            // Not a constant, give up
+            value.type = none;
+            break;
+          }
+        }
+        if (value.type != none) {
+          // Success! We can just apply the constant in the function, which makes
+          // the parameter value unused, which lets us remove it later.
+          Builder builder(*module);
+          func->body = builder.makeSequence(
+            builder.makeSetLocal(i, builder.makeConst(value)),
+            func->body
+          );
+          // Mark it as unused, which we know it now is (no point to
+          // re-scan just for that).
+          infoMap[name].unusedParams.insert(i);
+        }
+      }
+    }
+    // Track which functions we changed, and optimize them later if necessary.
+    std::unordered_set<Function*> changed;
+    // We now know which parameters are unused, and can potentially remove them.
+    for (auto& pair : allCalls) {
+      auto name = pair.first;
+      auto& calls = pair.second;
+      auto* func = module->getFunction(name);
+      auto numParams = func->getNumParams();
+      if (numParams == 0) continue;
+      // Iterate downwards, as we may remove more than one.
+      Index i = numParams - 1;
+      while (1) {
+        if (infoMap[name].unusedParams.has(i)) {
+          // Great, it's not used. Check if none of the calls has a param with side
+          // effects, as that would prevent us removing them (flattening before
+          // should have been done).
+          bool can = true;
+          for (auto* call : calls) {
+            auto* operand = call->operands[i];
+            if (EffectAnalyzer(runner->options, operand).hasSideEffects()) {
+              can = false;
+              break;
+            }
+          }
+          if (can) {
+            // Wonderful, nothing stands in our way! Do it.
+            // TODO: parallelize this?
+            removeParameter(func, i, calls);
+            changed.insert(func);
+          }
+        }
+        if (i == 0) break;
+        i--;
+      }
+    }
+    if (optimize && changed.size() > 0) {
+      OptUtils::optimizeAfterInlining(changed, module, runner);
+    }
+  }
+
+private:
+  void removeParameter(Function* func, Index i, std::vector<Call*> calls) {
+    // Clear the type, which is no longer accurate.
+    func->type = Name();
+    // It's cumbersome to adjust local names - TODO don't clear them?
+    Builder::clearLocalNames(func);
+    // Remove the parameter from the function. We must add a new local
+    // for uses of the parameter, but cannot make it use the same index
+    // (in general).
```

Comment:
Er, why not?

Reply:
Due to wasm declaring the params first, then the other locals. So if we had

`(func $x (param $a i32) (param $b i32) (local $c i32) ..`

Then if we remove the first param `$a`, which had index 0, there is no way to have a non-param local with that index, as `$b` must then have index 0.